experiments/yarn/src/main/scala/org/apache/gearpump/experiments/yarn/master/YarnApplicationMaster.scala

```diff
@@ -0,0 +1,338 @@
+import org.apache.hadoop.yarn.client.api.AMRMClient.ContainerRequest
+import org.apache.hadoop.yarn.client.api.async.impl.NMClientAsyncImpl
+import org.apache.hadoop.yarn.client.api.async.{AMRMClientAsync, NMClientAsync}
+import org.apache.hadoop.yarn.conf.YarnConfiguration
+import org.slf4j.Logger
+
+import scala.util.{Failure, Success, Try}
+
+object AmStates {
+
+  sealed trait Reason
+  case class Failed(throwable: Throwable) extends Reason
+  case object ShutdownRequest extends Reason
+  case object AllRequestedContainersCompleted extends Reason
+
+}
+
+object AmActorProtocol {
+  import AmStates._
+
+  case class AMRMClientAsyncStartup(status: Try[Boolean])
+  case class LaunchContainers(containers: List[Container])
+  case class LaunchWorkerContainers(containers: List[Container])
+  case class LaunchServiceContainer(containers: List[Container])
+  case class ContainerRequestMessage(memory: Int, vCores: Int)
+  case class RMHandlerDone(reason: Reason, rMHandlerContainerStats: RMHandlerContainerStats)
+  case class RMHandlerContainerStats(allocated: Int, completed: Int, failed: Int)
+  case class RegisterAMMessage(appHostName: String, appHostPort: Int, appTrackingUrl: String)
+  case class RegisterAppMasterResponse(response: RegisterApplicationMasterResponse)
+  case class AMStatusMessage(appStatus: FinalApplicationStatus, appMessage: String, appTrackingUrl: String)
+  case class ContainerStarted(containerId: ContainerId)
+
+}
+
+
+class YarnApplicationMaster(appConfig: AppConfig, yarnConf: YarnConfiguration,
+                            propsRMClient: Props,
+                            createRMCallbackHandler: (AppConfig, ActorRef) => ResourceManagerCallbackHandler,
+                            createNMClient: (NodeManagerCallbackHandler, YarnConfiguration) => NMClientAsync,
+                            createNodeManagerCallbackHandler: (ActorRef) => NodeManagerCallbackHandler,
+                            containerLaunchContextFactory: (String) => org.apache.hadoop.yarn.api.records.ContainerLaunchContext) extends Actor {
+
+  import AmActorProtocol._
+  import AmStates._
+
+  val LOG: Logger = LogUtil.getLogger(getClass)
+  val nodeManagerCallbackHandler = createNodeManagerCallbackHandler(self)
+  val nodeManagerClient: NMClientAsync = createNMClient(nodeManagerCallbackHandler, yarnConf)
+  val rmClientActor = context.actorOf(propsRMClient, "rmClient")
+
+  var masterContainers = Map.empty[ContainerId, (String, Int)]
+  val host = InetAddress.getLocalHost.getHostName
+  val servicesPort = appConfig.getEnv(SERVICES_PORT).toInt
+  val trackingURL = "http://" + host + ":" + servicesPort
+  var masterAddr: Option[HostPort] = None
+  var masterContainersStarted = 0
+  var workerContainersStarted = 0
+  var workerContainersRequested = 0
+  val version = appConfig.getEnv("version")
+
+  var servicesActor: Option[ActorRef] = None
+
+  override def preStart(): Unit = {
+    val rmCallbackHandler = createRMCallbackHandler(appConfig, self)
```

Comment:
I am not convinced that why putting RMCallbackHandler here.

I think you can put RMCallbackHandler in ResourceManagerClient. All message in RMCallbackHandler will be first sent to ResourceManagerClient, and then ResourceManagerClient forward back to YarnAppMaster if needed. 

It is strange to me that RMCallbackHandler accept an actorRef  of YarnAppMaster, it should accept actorRef of ResourceManagerClient instead.